launch week 5 blog posts (#8389)
highlight/highlight: blog-content/lw5-clickhouse-performance-optimization.md

issue a single `INSERT` command with many values. Though this worked to ensure inserts were large, inserts were not
atomic
resulting in duplicate data being inserted during worker reboots.

We opted to use the ClickHouse Kafka Connect Sink that implements batched writes and exactly-once semantics achieved
through ClickHouse Keeper. Check out a detailed overview of this feature here TODO.

To ensure our changes had the right effect, we monitored the Max Parts TODO graph in ClickHouse for the number of parts
waiting to be merged. Read more on how this works from ClickHouse here.

### Keeping Data in Wide Parts

CPU Usage during merges can be incurred by the conversion of parts from compact to wide. When data is inserted
compressed, it must be decompressed into the wide format for it to be merged. If inserting large batch sizes (100k+ rows
at 10MB+ data), a table level setting that helps is using min_rows_for_wide_part=0, min_bytes_for_wide_part=0 to make
sure that ClickHouse will keep inserted data as WIDE to avoid having to convert back and forth between the compact and
narrow format, since each conversion incurs a CPU cost.

### Optimizing Order By Granularity

> TODO what does the ORDER BY do

Author: still WIP

---
title: "Optimizing Clickhouse: 5 Tactics That Worked"
createdAt: 2024-04-30T09:00:00Z
readingTime: 17
authorFirstName: Vadim
authorLastName: Korolik
authorTitle: CTO @ Highlight
authorTwitter: 'https://twitter.com/vkorolik'
authorLinkedIn: 'https://www.linkedin.com/in/vkorolik/'
authorGithub: 'https://github.com/Vadman97'
authorWebsite: 'https://vadweb.us'
authorPFP: 'https://lh3.googleusercontent.com/a-/AOh14Gh1k7XsVMGxHMLJZ7qesyddqn1y4EKjfbodEYiY=s96-c'
tags: Launch Week 5
metaTitle: "Optimizing Clickhouse Cost + CPU: 5 Tactics That Worked"
---

Scaling an observability stack isn’t an easy problem. At Highlight.io, we ingest nearly 100 terabytes of data from our
customers every month. A large chunk of the volume goes to our ClickHouse cluster, handling our client’s Logs, Traces,
and Session metadata. Not only do we need to gracefully handle the volume and daily spikes in the traffic patterns, but
we also need to ensure that the data is ingested in an efficient format that is performant to query.

> 

Reviewer: ```suggestion
About a year ago, once we started onboarding larger customers, we started hitting scaling issues with our Clickhouse set up that warranted several "phases" of improvements to our query and ingest systems. This post goes through the initial realization that we had a challenge on our hands, and how we tackled it.
```

## The Incident

> As our data ingest has grown, we have needed to make improvements to our ClickHouse configuration to keep up and further
> improve query performance. For example, one day, we onboarded a large customer that sends close to 1 billion trace spans
> a day, adding another terabyte to our data ingestion. Beyond the total volume, the larger problem was in the number of

Reviewer: ```suggestion
One day, we onboarded a large customer that sent close to 1 billion trace spans
a day, adding another terabyte to our data ingestion. Beyond the total volume, the larger problem was in the number of
```

as the cluster created backpressure to reduce the overall CPU load. We also observed “too many parts” errors as
ClickHouse was unable to merge the data into fewer parts because of the CPU load. While the hotfix was increasing the
CPU allocated the cluster, we quickly had to find a long term solution to avoid the problem.

![ClickHouse CPU Wait](/images/blog/launch-week/5/clickhouse-cpu-wait.png)

Unlike traditional OTLP SQL, the way that data is written to a ClickHouse table affects query performance.
For instance, if data is written in smaller chunks, ClickHouse ends up creating more “parts” (in other words, files) to
store it. When querying, ClickHouse has to merge and possibly reorder the results from all relevant parts. Additionally,
ClickHouse will asynchronously merge data as it is initially ingested to get it into the right format for long term
storage, but this process consumes CPU on background threads that ultimately may decrease performance of concurrent
storage. Inserting data with the right configuration helps the merges that ClickHouse has to perform and can speed up
how much each merge takes.

> ## Optimizing CPU Usage: Our Strategies

Reviewer: ```suggestion
```

timestamps, ClickHouse will have to merge many parts as the different writes will land in different active parts. This,
in turn, will increase background CPU activity.

In our application (this will depend on your use case), we ensure that most writes are coming to
a few partitions, limiting the number of parts that are being merged. We also set a TTL on our tables to clear out old
data, which helps with parts remaining active. You can always check how many parts are active by running the following
SQL query:

```SQL
select table, count()
from clusterAllReplicas(default, system.parts)
where active
group by 1
order by 2 desc;
```

You can see what parts are active across different partitions to understand where writes are landing.

```SQL
select partition, level, count()
from clusterAllReplicas(default, system.parts)
where active and table = 'my_table'
group by 1, 2, 3
order by 1, 2, 3;
```

You can also observe all current merges to understand what tables / partitions are causing the most.

```SQL
select partition, sum(num_parts)
from clusterAllReplicas(default, system.merges)
where table = 'my_table'
group by 1
order by 2 desc;
```

> system drop parts

Reviewer: is this finished?
